### ISSUE TITLE
extend `prefer_const_declarations` to encourage const record literals

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#3626]
From [prefer_const_declarations](https://dart-lang.github.io/linter/lints/prefer_const_declarations.html),

> PREFER using const for const declarations.

> Const declarations are more hot-reload friendly and allow to use const constructors if an instantiation references this declaration.

**GOOD:**

```dart
const o = <int>[];

class A {
  static const o = <int>[];
}
```

**BAD:**

```dart
final o = const <int>[];

class A {
  static final o = const <int>[];
}
```

Should we extend this to include records?

/cc @munificent @goderbauer

**Comment by pq** [comment]
I'm inclined to this we _should_ extends this but could use some input from stakeholders.  (@goderbauer, @munificent: any reservations?)

**GOOD:**

```dart
const record = (number: 123, name: "Main", type: "Street");
```

**BAD:**

```dart
final tuple = ("first", 2, true);
```

**Comment by pq** [comment]
Implementation notes:

Prototyping, this doesn't work as expected w/ the LinterContext constant evaluator.

Specifically,


```dart
final tuple = ("first", 2, true);
```

produces a `CompileTimeErrorCode.MISSING_CONST_IN_LIST_LITERAL` error when passed to `evaluateConstant` (where I wouldn't expect it to).

/fyi @bwilkerson

**Comment by munificent** [comment]
Yeah, I think it's reasonable.
